recommender/recommender.py

```diff
@@ -235,98 +246,119 @@ def error_handler(self, error_msg, event, resource_id=None):
 
         result = {}
         result['id'] = resource_id
-        result['oldVotes'] = (self.recommendations[idx]['upvotes'] -
-                              self.recommendations[idx]['downvotes'])
-        if resource_id in self.downvoted_ids:
-            del self.downvoted_ids[self.downvoted_ids.index(resource_id)]
-            self.recommendations[idx]['downvotes'] -= 1
-            result['newVotes'] = (self.recommendations[idx]['upvotes'] -
-                                  self.recommendations[idx]['downvotes'])
+        is_event_upvote = (data['event'] == 'recommender_upvote')
+        result['oldVotes'] = (self.recommendations[resource_id]['upvotes'] -
+                              self.recommendations[resource_id]['downvotes'])
+
+        if is_event_upvote:
+            if resource_id in self.upvoted_ids:
+                del self.upvoted_ids[self.upvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['upvotes'] -= 1
+                result['newVotes'] = (self.recommendations[resource_id]['upvotes'] -
+                                      self.recommendations[resource_id]['downvotes'])
+        else:
+            if resource_id in self.downvoted_ids:
+                del self.downvoted_ids[self.downvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['downvotes'] -= 1
+                result['newVotes'] = (self.recommendations[resource_id]['upvotes'] -
+                                      self.recommendations[resource_id]['downvotes'])
+
+        if 'newVotes' in result:
             result['Success'] = True
-            tracker.emit('recommender_downvote', result)
+            tracker.emit(data['event'], result)
             return result
 
-        if resource_id in self.upvoted_ids:
-            del self.upvoted_ids[self.upvoted_ids.index(resource_id)]
-            self.recommendations[idx]['upvotes'] -= 1
-            result['toggle'] = True
-        self.downvoted_ids.append(resource_id)
-        self.recommendations[idx]['downvotes'] += 1
-        result['newVotes'] = (self.recommendations[idx]['upvotes'] -
-                              self.recommendations[idx]['downvotes'])
+        if is_event_upvote:
+            if resource_id in self.downvoted_ids:
+                del self.downvoted_ids[self.downvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['downvotes'] -= 1
+                result['toggle'] = True
+            self.upvoted_ids.append(resource_id)
+            self.recommendations[resource_id]['upvotes'] += 1
+        else:
+            if resource_id in self.upvoted_ids:
+                del self.upvoted_ids[self.upvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['upvotes'] -= 1
+                result['toggle'] = True
+            self.downvoted_ids.append(resource_id)
+            self.recommendations[resource_id]['downvotes'] += 1
```

Comment:
I don't understand what this block of code is doing, and it looks like the logic is fairly repetitive. Please add some comments explaining what you're doing. The process of adding comments may help you to refactor as well, to avoid duplicated code.